include/dav1d/picture.rs

```diff
@@ -102,44 +109,184 @@ pub struct Dav1dPicture {
+    #[inline(always)] // So the caller can see the real alignment.
     unsafe fn as_mut_ptr(ptr: *mut Self) -> *mut Self::Target {
         // SAFETY: Safe to dereference by unsafe preconditions.
         // Since we don't store any `&mut`s, just a raw ptr, we can have a `&Self`.
         let this = unsafe { &*ptr };
-        this.ptr.cast().as_ptr()
+
+        // Assume this so that the compiler knows `ptr` is aligned.
+        // Normally we'd store this as a slice so the compiler would know,
+        // but since it's a ptr due to `DisjointMut`, we explicitly assume it here.
+        // SAFETY: We already checked this in `Self::new`.
+        unsafe { assume(this.ptr.is_aligned()) };
+
+        this.ptr.cast::<u8>().as_ptr()
     }
 
     fn len(&self) -> usize {
-        self.len
+        self.len * mem::size_of::<AlignedPixelChunk>()
+    }
+}
+
+pub struct Rav1dPictureDataComponent(DisjointMut<Rav1dPictureDataComponentInner>);
+
+impl Rav1dPictureDataComponent {
+    fn chunk_len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Length in number of pixels.
+    pub fn len<BD: BitDepth>(&self) -> usize {
+        self.chunk_len() * (mem::size_of::<AlignedPixelChunk>() / mem::size_of::<BD::Pixel>())
+    }
+
+    fn chunk_stride(&self) -> isize {
+        // SAFETY: We're only accessing the `stride` fields, not `ptr`.
+        unsafe { (*self.0.inner()).stride }
+    }
+
+    /// Stride in number of pixels.
+    pub fn stride<BD: BitDepth>(&self) -> isize {
+        self.chunk_stride()
+            * (mem::size_of::<AlignedPixelChunk>() / mem::size_of::<BD::Pixel>()) as isize
+    }
+
+    /// Strided ptr to chunks.
+    fn as_chunk_mut_ptr(&self) -> *mut AlignedPixelChunk {
+        let ptr = self.0.as_mut_ptr().cast::<AlignedPixelChunk>();
+        let stride = self.chunk_stride();
+        if stride < 0 {
+            // SAFETY: This puts `ptr` one past the end of the data.
+            let ptr = unsafe { ptr.add(self.chunk_len()) };
+            // SAFETY: `stride` is negative and `-stride < len`, so this should stay in bounds.
+            let ptr = unsafe { ptr.offset(stride) };
+            ptr
+        } else {
+            ptr
+        }
+    }
+
+    /// Strided ptr to pixels.
+    pub fn as_mut_ptr<BD: BitDepth>(&self) -> *mut BD::Pixel {
+        self.as_chunk_mut_ptr().cast()
+    }
+
+    /// Strided ptr to pixels.
+    pub fn as_ptr<BD: BitDepth>(&self) -> *const BD::Pixel {
+        self.as_mut_ptr::<BD>().cast_const()
+    }
+
+    fn as_dav1d(&self) -> Option<NonNull<c_void>> {
+        if self.chunk_len() == 0 {
+            None
+        } else {
+            NonNull::new(self.as_chunk_mut_ptr().cast())
+        }
+    }
+
+    pub fn copy_from(&self, src: &Self) {
+        let dst = &mut *self.0.index_mut(..);
+        let src = &*src.0.index(..);
+        dst.clone_from_slice(src);
+    }
+
+    #[inline(always)] // Need to inline to see bounds checks in order to potentially elide them.
```

Comment:
Some of them can be, but on at least one of them it wasn't inlining when I did just `#[inline]`, even in release.  I can check again, though, since I made the error path `#[inline(never)]` now.

Reply:
I'd still just use `#[inline]`, that means the function is large and perhaps shouldn't be inlined.

Reply:
> I'd still just use `#[inline]`, that means the function is large and perhaps shouldn't be inlined.

But we do know better than LLVM here.  Its size heuristic isn't as good as us knowing that there is a ton of optimization info lying on both sides of the function call.

Reply:
It seems `#[inline]` now is sufficient, so I'll switch to that.

Reply:
Changed in [`9eb1d74..4e786d4`](https://github.com/memorysafety/rav1d/compare/9eb1d74618455244c89efb6f3f6f2cfc9d24f752..4e786d4d73375dab5966a51ea17e120bff6951ae).